library/array.lisp

```diff
@@ -0,0 +1,110 @@
+(coalton-library/utils:defstdlib-package #:coalton-library/array
+  (:use
+   #:coalton
+   #:coalton-library/classes)
+  (:local-nicknames
+   (#:c #:coalton-library/math/complex)
+   (#:a #:coalton-library/math/arith))
+  (:export
+   #:Array
+   #:make
+   #:size
+   #:aref
+   #:set!
+
+   #:SingleFloats
+   #:DoubleFloats
+   #:ComplexSingleFloats
+   #:ComplexDoubleFloats))
+
+(in-package #:coalton-library/array)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+;;; Class Implementation
+(coalton-toplevel
+  ;; The decision to use a functional dependency was an ergonomic one,
+  ;; not a technical one. If we did not establish this dependency, we
+  ;; would win flexibility (a single storage type can store multiple
+  ;; data types), but we would lose a lot of practical ergonomics
+  ;; (e.g., we would need Proxy types, calls to size would need to be
+  ;; disambiguated, etc.).
+  ;;
+  ;; While we lose some flexibility, we still retain some. For
+  ;; instance, we can have multiple storage types for double floats
+  ;; (think: Lisp arrays, C arrays, GPU arrays, etc.).
+  (define-class (Array :f :t (:f -> :t))
+    "Establishes that `:f` can (only) store elements of type `:t`. The **storage type** `:f` implies the **stored type** `:t`.
+
+The storage is expected to be sequential and efficient in random access."
+    (make (UFix -> :t -> :f))
+    (size (:f -> UFix))
+    (aref (:f -> UFix -> :t))
+    (set! (:f -> UFix -> :t -> Unit))))
+
+;;; Built-In Storage Types
+(cl:defmacro %define-unboxed-array (cl-type ct-type coalton-array-type)
+  `(coalton-toplevel
+     (repr :native (cl:simple-array ,cl-type (cl:*)))
+     (define-type ,coalton-array-type)
+
+     (define-instance (Array ,coalton-array-type ,ct-type)
+       (define (make n x)
+         (lisp ,coalton-array-type (n x)
+           (cl:make-array n :element-type ',cl-type :initial-element x)))
+       (define (size a)
+         (lisp UFix (a)
+           (cl:length a)))
+       (define (aref a n)
+         (lisp ,ct-type (a n)
+           (cl:aref a n)))
+       (define (set! a n x)
+         (lisp Unit (a n x)
+           (cl:setf (cl:aref a n) x)
+           Unit)))))
+
+(%define-unboxed-array cl:single-float Single-Float SingleFloats)
```

Comment:
This seems like a really useful type, and I could see myself reworking a lot of stuff with this in mind.

I wonder if there might be utility in having a macro/function like `define-type-with-array` that would automatically add an array type when you define the new type. I know that's probably a later stage polish.

Reply:
Wouldn't unboxed-arrays be limited by the underlying lisp implementation?

Reply:
@Izaakwltn Only some kinds of base types might benefit from an array type in Lisp. Other types might need more finesse, like, for example, an unboxed array of RGB values, which might require working with foreign memory.

Moreover, it's possible that we define `(Vector :t)` as a kind of storage of `:t`, so that there's at least one generically useful parametric storage type for all types `:t`, even if it's not as optimized as `DoubleFloats` and friends.